### PR TITLE
Add Participationmanager

### PR DESCRIPTION
**Pull request by HellPat**
I wanted to fully decouple the Engine from storage.
Since Storage is now general purpose there is nobody handling the name of cookie or session prefixes at the moment.

That's where ParticipationManager comes in. We can write and read from the ParticipationManager
and inject it to the Engine.

The constructor of the engine could then look like this. (or we add it to the eventDispatcher.
But imo this should be a hard dependency on the class.

``` php
public function __construct(
   DispatcherInterface $dispatcher,
   ParticipationManagerInterface $participationManager
)
```

what do you think of adding another layer of abstraction here?
I think that it's good to have this in place as we can also use it in our FrontendHelpers later.

Could one of you take this?

### CODE REVIEW

**Comment by waltertamboer** [comment]
What do you think about this?

``` php
<?php

class StorageListener
{
    public funciton __construct(StorageInterface $storage)
    {
        // ...
    }

    public function __invoke($options)
    {
        $test = $options->getParam('test');
        $variant = $options->getParam('variant');

        $this->storage->doStorage($test, $variant);
    }
}

class Engine
{
    public function start()
    {
        $this->dispatcher->addListener('execute', [$this, 'onChooseVariant'], 0);

        foreach ($this->tests as $test) {
            $event = new ...();
            $event->setParam('test', $test);

            $this->dispatcher->dispatch('execute', $event);
        }
    }

    // Or maybe move this method to it's own listener class.
    public function onChooseVariant($event)
    {
        $variant = ...;

        $event->setParam('variant', $variant);
    }
}

$cookieStorage = new CookieStorage();
$otherStorage = new OtherStorage();

$dispatcher = new ...();

// A minus priority because we want to store _after_ the test has been executed.
$dispatcher->addListener('execute', new StorageListener($cookieStorage), -1);
$dispatcher->addListener('execute', new StorageListener($otherStorage), -1);

$engine = new Engine($dispatcher);
$engine->start();
```

**Comment by waltertamboer** [comment]
I've added two storage listeners to show the flexibility of this concept.

**Comment by HellPat** [comment]
I thought about that, too but was not sure. Engine will not be working if no StorageListener is attached.
I fear ab bit for people forgetting to add it.

**Comment by waltertamboer** [comment]
In the subpackages (zf2 module, symfony bundle, etc.) we could force that of course. But I see your point and have no good arguments against them to be honest (except for the one I just mentioned).

**Comment by HellPat** [comment]
I'd like to go with constructor injection as we need at least one implementation here.
If something has special needs (like analytics) this implementation can use our events.
Is it okay from your point of view?

**Comment by waltertamboer** [comment]
Yes, let's try it out.

**Comment by HellPat** [comment]
I implemented ParticipationManager. Could you have a look?

I also found some inconsitstencies between storage providers.
We should remove the namespacing feature from storage (since it's general purpose now) and let the participationManager handle this. What do you think? when i switch the different storages in my testcase for participationManager i get errors, too. Could somebody catch that up?

File: src/Participation/Manager.php [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    public function __construct(StorageInterface $storage, $prefix = 'phpab_')
+    {
+        $this->storage = $storage;
+        $this->prefix = $prefix;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function participates($test, $variant = null)
```
Comment (HellPat) [inline src/Participation/Manager.php:27]: We have to lower cyclomatic complexity here somhow. Suggestions?

Comment (waltertamboer) [inline src/Participation/Manager.php:27]: I don't see the problem at this point. Would leave it as is till we have different use cases.

File: src/Participation/Manager.php [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    {
+        return $this->prefix.$identifier;
+    }
+
+    /**
+     * @param mixed $subject The subject to get the identifier for
+     *
+     * @return string The identifier of the item
+     */
+    private function getIdentifier($subject)
```
Comment (waltertamboer) [inline src/Participation/Manager.php]: I think this is a dangerous method that does not offer any additional behaviour. Why not just call the getIdentifier method on the objects themselves?

**Comment by waltertamboer** [comment]
Why would you handle the namespacing in the participation manager? Namespacing might not be relevant for all storage providers. For cookies and sessions it is because you don't want to have conflicts with the application session and cookies. I think the participation manager should just work with the raw data.

What conflicts did you notice?

**Comment by HellPat** [comment]
Try switching Runtime/Session/Cookie in `ManagerTest:L15`

File: src/Participation/Manager.php (hunk 2 above)
Comment (HellPat) [inline src/Participation/Manager.php]: I wanted to allow sting or object. I have to check for instanceof and it seemed like copy&paste between participate and participates

**Pull request merged**